plugins/api/src/replica_open.cpp

```diff
@@ -0,0 +1,158 @@
+#include "api_plugin_number.h"
+#include "dataObjInpOut.h"
+#include "rodsDef.h"
+#include "rcConnect.h"
+#include "rodsErrorTable.h"
+#include "rodsPackInstruct.h"
+#include "rcMisc.h"
+#include "client_api_whitelist.hpp"
+
+#include "apiHandler.hpp"
+
+#include <functional>
+
+#ifdef RODS_SERVER
+
+//
+// Server-side Implementation
+//
+
+#include "objDesc.hpp"
+#include "irods_stacktrace.hpp"
+#include "irods_server_api_call.hpp"
+#include "irods_re_serialization.hpp"
+#include "rsDataObjOpen.hpp"
+#include "rsDataObjClose.hpp"
+#include "rs_get_file_descriptor_info.hpp"
+#include "irods_logger.hpp"
+
+#include <json.hpp>
+
+#include <string>
+
+namespace
+{
+    using log = irods::experimental::log;
+
+    //
+    // Function Prototypes
+    //
+
+    auto call_replica_open(irods::api_entry*, rsComm_t*, dataObjInp_t*, bytesBuf_t**) -> int;
+
+    auto rs_replica_open(rsComm_t*, dataObjInp_t*, bytesBuf_t**) -> int;
+
+    //
+    // Function Implementations
+    //
+
+    auto call_replica_open(irods::api_entry* _api, rsComm_t* _comm, dataObjInp_t* _input, bytesBuf_t** _output) -> int
+    {
+        return _api->call_handler<dataObjInp_t*, bytesBuf_t**>(_comm, _input, _output);
+    }
+
+    auto rs_replica_open(rsComm_t* _comm, dataObjInp_t* _input, bytesBuf_t** _output) -> int
+    {
+        if (!_input || !_output) {
+            log::api::error("Invalid input detect.");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        log::api::trace("Opening data object ...");
+
+        const auto fd = rsDataObjOpen(_comm, _input);
+
+        log::api::trace("Opened file descriptor = {}", fd);
+
+        if (fd < 3) {
+            log::api::error("Could not open replica [error_code={}]", fd);
+            return fd;
+        }
+
+        log::api::trace("Constructing JSON ...");
+
+        const auto json_input = nlohmann::json{{"fd", fd}}.dump();
+        char* json_output{};
+
+        log::api::trace("JSON = {}", json_input);
+        log::api::trace("Retrieving file descriptor information ...");
+
+        if (const auto ec = rs_get_file_descriptor_info(_comm, json_input.data(), &json_output); ec != 0) {
```

Comment:
this function needs to support redirect because the actual info is in the remote computer.

Reply:
done.